### PR TITLE
Add option to print line info on verification

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add option to suppress line info.

Resolves: #308 

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

**Review by dthaler — COMMENTED** [review]
Maybe also add a comment-line option to check.cpp to set this option.

**Review by dthaler — COMMENTED** [review]
The PR title says it's an option to "suppress" line info, but the actual option has opposite parity.

Regardless, please update config.cpp to initialize the default value.

**Review by dthaler — APPROVED** [review]

**Review by dthaler — COMMENTED** [review]
Still missing a change to src/config.cpp to explicitly initialize the field (it will default to false but for consistency all the other fields are initialized explicitly so this one should be too)

File: src/asm_ostream.cpp [hunk 1]
```diff
@@ -485,3 +486,9 @@ std::ostream& operator<<(std::ostream& o, const cfg_t& cfg) {
     }
     return o;
 }
+
+std::ostream& operator<<(std::ostream& os, const btf_line_info_t& line_info) {
+    os << "; " << line_info.file_name.c_str() << ":" << line_info.line_number << "\n";
+    os << "; " << line_info.source_line.c_str() << "\n";
+    return os;
+}
```
Comment (elazarg) [inline src/asm_ostream.cpp:494]: Missing line ending

File: src/asm_ostream.cpp [hunk 2]
```diff
@@ -357,7 +357,8 @@ auto get_labels(const InstructionSeq& insts) {
     return pc_of_label;
 }
 
-void print(const InstructionSeq& insts, std::ostream& out, std::optional<const label_t> label_to_print) {
+void print(const InstructionSeq& insts, std::ostream& out, std::optional<const label_t> label_to_print,
+           std::function<void(std::ostream&, const btf_line_info_t&)> line_info_formatter) {
```
Comment (elazarg) [inline src/asm_ostream.cpp:361]: The new parameter is unused. What am I missing?

File: src/asm_ostream.cpp [hunk 3]
```diff
@@ -485,3 +486,9 @@ std::ostream& operator<<(std::ostream& o, const cfg_t& cfg) {
     }
     return o;
 }
+
+std::ostream& operator<<(std::ostream& os, const btf_line_info_t& line_info) {
+    os << "; " << line_info.file_name.c_str() << ":" << line_info.line_number << "\n";
+    os << "; " << line_info.source_line.c_str() << "\n";
```
Comment (elazarg) [inline src/asm_ostream.cpp:492]: Why the access to `c_str()`?

File: src/main/check.cpp [hunk 4]
```diff
@@ -57,6 +57,7 @@ int main(int argc, char** argv) {
     app.add_flag("-s", ebpf_verifier_options.strict, "Apply additional checks that would cause runtime failures");
     app.add_flag("-v", verbose, "Print both invariants and failures");
     app.add_flag("--no-simplify", ebpf_verifier_options.no_simplify, "Do not simplify");
+    app.add_flag("-p", ebpf_verifier_options.print_line_info, "Print line information");
```
Comment (elazarg) [inline src/main/check.cpp:60]: I'm not sure the letter is indicative. Maybe `n`? Or simply a longer flag such as `--line-info`.

File: src/crab_verifier.cpp [hunk 5]
```diff
@@ -199,7 +198,7 @@ ebpf_analyze_program_for_test(std::ostream& os, const InstructionSeq& prog, cons
     cfg_t cfg = prepare_cfg(prog, info, !no_simplify, false);
     auto [pre_invariants, post_invariants] = crab::run_forward_analyzer(cfg, entry_inv, check_termination);
     checks_db report = generate_report(cfg, pre_invariants, post_invariants);
-    print_report(os, report, prog);
+    print_report(os, report, prog, [](std::ostream& out, const btf_line_info_t& info) { out << info; });
```
Comment (elazarg) [inline src/crab_verifier.cpp:201]: This lambda is also defined below. I think it's cleaner to have both as static functions outside.

But why use a formatter if the only decision is whether or not to print? Again it feels like I'm missing something. Are you planning to add more line-info formatting?

**Review by elazarg — COMMENTED** [review]
I do not entirely understand the use of callbacks.

File: src/asm_ostream.cpp (hunk 2 above)
Comment (Alan-Jowett) [reply src/asm_ostream.cpp]: Simplified this to just accept a bool.

**Pull request merged**